### PR TITLE
optimising intersection of trajectory with RegularGrid

### PR DESCRIPTION
**Pull request by andy-beer**
closes issue #395

### CODE REVIEW

**Comment by andy-beer** [comment]
@maxm17 This code is ready for review (the pipeline failure is a toml issue). I don't expect you to follow all the workings of the changes but please have a look over it in case you spot any obvious mistakes.  The change re-works how the RegularGrid class works with a Crs in order to preserve the origin of the grid.

File: resqpy/grid/grid.py [hunk 1]
```diff
@@ -553,14 +553,16 @@ def z_units(self):
             return None
         return rqet.find_tag(crs_root, 'VerticalUom').text
 
-    def skin(self, use_single_layer_tactics = False):
+    def skin(self, use_single_layer_tactics = False, is_regular = False):
         """Returns a GridSkin composite surface object reoresenting the outer surface of the grid."""
 
         import resqpy.grid_surface as rqgs
```
Comment (maxm17) [inline resqpy/grid/grid.py:559]: I know it wasn't added here but the imports should be at the top of the file

File: resqpy/grid/xyz.py [hunk 2]
```diff
@@ -179,13 +178,11 @@ def global_to_local_crs(grid,
     # todo: check resqml definition for order of rotation and translation and apply rotation if not zero
     if global_xy_units is not None:
         crs_xy_units_text = rqet.find_tag(crs_root, 'ProjectedUom').text
-        if crs_xy_units_text in ['ft', 'm']:  # todo: else raise exception
```
Comment (maxm17) [inline resqpy/grid/xyz.py:182]: I've already made changes to global_to_local_crs() but I can resolve the conflicts

File: resqpy/grid_surface.py [hunk 3]
```diff
@@ -42,10 +42,42 @@ def __init__(self, grid, quad_triangles = True, use_single_layer_tactics = True)
...
+                if grid.block_dxyz_dkji[3 - axis, axis] < 0.0:
+                    xyz_box[:, axis] = (xyz_box[1, axis], xyz_box[0, axis])
+            min_x, min_y, min_z = xyz_box[0]
+            max_x, max_y, max_z = xyz_box[1]
+            top_surf = rqs.Surface(grid.model)
+            top_surf.set_to_horizontal_plane(0.0, xyz_box)
+            base_surf = rqs.Surface(grid.model)
+            base_surf.set_to_horizontal_plane(grid.nk * grid.block_dxyz_dkji[0, 2], xyz_box)
+            j_minus_surf = rqs.Surface(grid.model)
+            corners = np.array([(min_x, min_y, min_z), (max_x, min_y, min_z), (min_x, min_y, max_z),
```
Comment (maxm17) [inline resqpy/grid_surface.py:63]: Could be converted to a loop but not necessary

**Review by maxm17 — COMMENTED** [review]
Just a few things that could be changed

File: tests/unit_tests/well/test_well_object_funcs.py [hunk 4]
```diff
@@ -134,25 +134,28 @@ def test_add_logs_from_cellio_file(example_model_and_crs):
...
     well_name = 'Coconut'
     source_dataframe = pd.DataFrame({
         'MD': mds,
-        'X': [25, 50, 75, 100, 100],
-        'Y': [25, -50, -75, -100, -100],
+        'X': [location[0], 25.0, 50, 75, 100, 100],
+        'Y': [location[1], 25.0, -50, -75, -100, -100],
         'Z': zs,
-        'WELL': ['Coconut', 'Coconut', 'Coconut', 'Coconut', 'Coconut']
+        'WELL': [well_name, well_name, well_name, well_name, well_name, well_name]
```
[the remark this answers is not in the document]
Comment (andy-beer) [reply tests/unit_tests/well/test_well_object_funcs.py:158]: I thought there was some way to do that – tried it inside the [] but just ended up with one long string!  Thanks for the hint.

**Comment by andy-beer** [comment]
Please could you do the conflict resolution, @maxm17  – thanks

**Pull request merged**